proposals/0000-add-support-for-display-p3-colors.md

```diff
@@ -0,0 +1,202 @@
+
+```java
+if (value instanceof ReadableMap) {
+  ReadableMap map = (ReadableMap) value;
+  float red = (float) map.getDouble("red");
+  float green = (float) map.getDouble("green");
+  float blue = (float) map.getDouble("blue");
+  float alpha = (float) map.getDouble("alpha");
+  int colorSpace = map.getInt("colorSpace")
+
+  if (colorSpace === 1) {
+    @ColorLong long p3 = pack(red, green, blue, alpha, ColorSpace.Named.DISPLAY_P3);
+    return Color.valueOf(p3);
+  } else {
+    // ...
+  }
+}
+```
+
+3. Determine how best to utilize wide gamut color in Android. It's not clear if [Android View](<https://developer.android.com/reference/android/view/View#setBackgroundColor(int)>)'s actually support it directly. It is supported by [Paint](<https://developer.android.com/reference/android/graphics/Paint#setColor(long)>) so perhaps it's possible via [setLayerPaint](<https://developer.android.com/reference/android/view/View#setLayerPaint(android.graphics.Paint)>)?
+
+### Docs
+
+Update the [Color Reference](https://github.com/facebook/react-native-website/blob/main/docs/colors.md) to document color function usage in React Native.
+
+## Drawbacks
+
+There should be no breaking changes for users but the color implementation will be a bit more complicated than before. On iOS colors specified using existing methods will still use the same code paths and UIColor constructor. On Android colors specified using existing methods will still default to sRGB and only if DisplayP3 is specified will they be packed into a ColorLong.
+
+## Alternatives
+
+1. **Alternatives to CSS Color Module Level 4:** <br/> We could use a different standard for displaying wide gamut color. For example we could follow the [Color.js](https://github.com/color-js/color.js) API: `new Color({space: "p3", coords: [0, 1, 0], alpha: .9})`. However this deviates from the existing color implementation which mostly implements [CSS Color Module Level 4](https://www.w3.org/TR/css-color-4/#color-function) and would be surprising to existing React Native developers.
+
+2. **Utilizing Swizzling on iOS:** <br/> As an alternative to the proposed iOS changes, swizzling could be considered similar to [RN #41517](https://github.com/facebook/react-native/issues/41517). This has a major downside of being an all or nothing breaking change forcing all React Native users into the DisplayP3 color space as well as being a more confusing and less maintainable solution.
+
+3. **Full switch to DisplayP3:** <br/> In a similar vein we could just update the constructors directly without preserving existing behavior. While this would be a simpler implementation it would be a breaking change. Any existing visual regression tests would likely break and React Native users would now be required to specify all colors in a wider gamut.
```

Comment:
> What do you think about a way to optionally fully opt-in to use DisplayP3?

I can definitely see the utility of this. I'm not really sure what the API for that should look like though. AFAIK the APIs on Android, iOS, and web all only allow specifying color space per instance. Since color properties already mostly implement CSS Color Module Level 4 we were mainly planning on extending that further with this proposed API.

Reply:
I agree that it could be useful. 
The way I think this should be implemented is by factoring out the constructor functions in a single factory method so we can decide which initializer we want to use as default. At that point, we can specify a static feature flag that controls the default behavior.

For Android things are a little bit more convoluted because Android uses `int` to represent sRGB colors and `long` to represent WideGamut colors. So, not only we need to update how a color is created but we also have to overload several methods to achieve what we need.

But it is probably doable.

@ryanlntn could you update the proposal to add a feature flag that will control the default init behavior?